plugin/advisor/catalog/walk_through.go

```diff
@@ -0,0 +1,427 @@
+		// we do not deal with FOREIGN KEY constraints
+	case tidbast.ConstraintFulltext:
+		keyList, err := t.validateAndGetKeyStringList(constraint.Keys, false /* primary */)
+		if err != nil {
+			return err
+		}
+		if err := t.createIndex(constraint.Name, keyList, false /* unique */, FullTextName); err != nil {
+			return err
+		}
+	case tidbast.ConstraintCheck:
+		// we do not deal with CHECK constraints
+	}
+
+	return nil
+}
+
+func (t *tableState) validateAndGetKeyStringList(keyList []*tidbast.IndexPartSpecification, primary bool) ([]string, error) {
+	var res []string
+	for _, key := range keyList {
+		if key.Expr != nil {
+			str, err := restoreNode(key, format.DefaultRestoreFlags)
+			if err != nil {
+				return nil, err
+			}
+			res = append(res, str)
+		} else {
+			columnName := key.Column.Name.O
+			column, exists := t.columnSet[columnName]
+			if !exists {
+				return nil, &WalkThroughError{
+					Type:    ErrorTypeColumnNotExists,
+					Content: fmt.Sprintf("Column `%s` in table `%s` not exists", columnName, t.name),
+				}
+			}
+			if primary {
+				column.nullable = false
+			}
+			res = append(res, columnName)
+		}
+	}
+	return res, nil
+}
+
+func (t *tableState) createColumn(column *tidbast.ColumnDef, position int) error {
+	if _, exists := t.columnSet[column.Name.Name.O]; exists {
+		return &WalkThroughError{
+			Type:    ErrorTypeColumnExists,
+			Content: fmt.Sprintf("Column `%s` exists in table `%s`", column.Name.Name.O, t.name),
+		}
+	}
+
+	col := &columnState{
+		name:         column.Name.Name.O,
+		position:     position,
+		columnType:   column.Tp.CompactStr(),
+		characterSet: column.Tp.GetCharset(),
+		collation:    column.Tp.GetCollate(),
+		nullable:     true,
+	}
+
+	for _, option := range column.Options {
+		switch option.Tp {
+		case tidbast.ColumnOptionPrimaryKey:
+			col.nullable = false
+			if err := t.createPrimaryKey([]string{col.name}, model.IndexTypeBtree.String()); err != nil {
+				return err
+			}
+		case tidbast.ColumnOptionNotNull:
+			col.nullable = false
+		case tidbast.ColumnOptionAutoIncrement:
+			// we do not deal with AUTO-INCREMENT
+		case tidbast.ColumnOptionDefaultValue:
+			defaultValue, err := restoreNode(option.Expr, format.RestoreStringWithoutCharset)
+			if err != nil {
+				return err
+			}
+			col.defaultValue = &defaultValue
+		case tidbast.ColumnOptionUniqKey:
+			if err := t.createIndex("", []string{col.name}, true /* unique */, model.IndexTypeBtree.String()); err != nil {
+				return err
+			}
+		case tidbast.ColumnOptionNull:
+			col.nullable = true
+		case tidbast.ColumnOptionOnUpdate:
+			// we do not deal with ON UPDATE
+		case tidbast.ColumnOptionComment:
+			comment, err := restoreNode(option.Expr, format.RestoreStringWithoutCharset)
+			if err != nil {
+				return err
+			}
+			col.comment = comment
+		case tidbast.ColumnOptionGenerated:
+			// we do not deal with GENERATED ALWAYS AS
+		case tidbast.ColumnOptionReference:
+			// MySQL will ignore the inline REFERENCE
```

Comment:
I think that was a fault in MySQL design but we can be more restricted as most users use innodb engine.

Reply:
I'll validate the foreign key in the following PR.